Add section to docs on duplicate messages in changefeeds (#17442)
cockroachdb/docs: src/current/v23.1/changefeed-messages.md

If you require `resolved` message frequency under `30s`, then you **must** set the [`min_checkpoint_frequency`](create-changefeed.html#min-checkpoint-frequency) option to at least the desired `resolved` frequency. This is because `resolved` messages will not be emitted more frequently than `min_checkpoint_frequency`, but may be emitted less frequently.
{{site.data.alerts.end}}

## Duplicate messages

Under some circumstances, changefeeds will emit duplicate messages to ensure the sink is receiving each message at least once. The following can cause duplicate messages:

> - The changefeed job [encounters an error](#changefeed-encounters-an-error).

Reviewer: Or is paused manually, as I think you cover elsewhere.

Author: I have added again here.

> - A node in the cluster [restarts](#node-restarts).

Author: Added that this could also be failing.

> - The changefeed job has the [`min_checkpoint_frequency` option set](#min_checkpoint_frequency-option).

Author: I have adjusted the language slightly, introducing the bullet points, and I have adapted the sentence in the bullet, do you think this is enough? Or could it use more clarification.

> - A target table undergoes a schema change. Refer to [Schema changes](#schema-changes) for detail on duplicates in this case.

Reviewer: Up to you, but it might be worth emphasizing here (for those who might make assumptions and not follow the link), that this can include emitting the entire table, not just recent messages.

Author: Yes, we should do that. I have included an additional sentence to that effect.

<img src="{{ 'images/v23.1/changefeed-duplicate-messages-emit.png' | relative_url }}" alt="How checkpoints will re-emit messages when a changefeed pauses. The changefeed returns to the last checkpoint and potentially sends duplicate messages." style="border:0px solid #eee;max-width:100%" />

### Changefeed encounters an error

By default, changefeeds treat errors as [retryable except for some specific terminal errors](monitor-and-debug-changefeeds.html#changefeed-retry-errors). When a changefeed encounters a retryable or non-retryable error, the job will pause until a successful retry or you resume the job once the error is solved. This can cause duplicate messages at the sink as the changefeed returns to the last checkpoint.

We recommend monitoring for changefeed retry errors and failures. Refer to the [Monitor and Debug Changefeeds](monitor-and-debug-changefeeds.html#recommended-changefeed-metrics-to-track) page.

{{site.data.alerts.callout_info}}

> [Kafka's](changefeed-sinks.html#kafka) batching behavior can increase the number of duplicate messages. For example, if Kafka receives a batch of `N` messages and successfully saves `N-1` of them, the changefeed job only knows that the batch failed, not which message failed to commit. As a result, the changefeed job will resend the full batch of messages, which means all but one of the messages are duplicates. Reducing the Kafka batch size with [`kafka_config_sink`](changefeed-sinks.html#kafka-sink-configuration) may help to reduce the number of duplicate messages at the sink.

Reviewer: `kafka_sink_config`, not `kafka_config_sink`.

Author: yikes! Thanks for the catch

> [Kafka's](changefeed-sinks.html#kafka) batching behavior can increase the number of duplicate messages. For example, if Kafka receives a batch of `N` messages and successfully saves `N-1` of them, the changefeed job only knows that the batch failed, not which message failed to commit. As a result, the changefeed job will resend the full batch of messages, which means all but one of the messages are duplicates. Reducing the Kafka batch size with [`kafka_config_sink`](changefeed-sinks.html#kafka-sink-configuration) may help to reduce the number of duplicate messages at the sink.

Reviewer: We observed this with Kafka, but it could happen with any sink that supports client-side batching. Which might just be Kafka, webhook, and cloud storage.

Author: OK! Yes, good point, I have reworked this note to make it more general with Kafka being and "example" and directing users to check out config options on the sinks page. I didn't specifically name the other sinks, because it concerns me that this note would be dated when other sinks might include batching behavior in the future. Does the update work?

Refer to the [Changefeed Sinks](changefeed-sinks.html) page for details on sink batching configuration.
{{site.data.alerts.end}}

### Node restarts

When a node restarts, the changefeed will emit duplicates since the last checkpoint. During a rolling restart of nodes, a changefeed can fall behind as it tries to catch up during each node restart. For example, as part of a rolling upgrade or cluster maintenance, a node may [drain](node-shutdown.html) every 5 minutes and the changefeed job checkpoints every 5 minutes.

To prevent the changefeed from falling too far behind, [pause](create-and-configure-changefeeds.html#configuring-all-changefeeds) changefeed jobs before performing rolling node restarts.

### `min_checkpoint_frequency` option

> The `min_checkpoint_frequency` option controls how often nodes flush their progress to the coordinating changefeed node. Therefore, changefeeds will wait for at least the `min_checkpoint_frequency` duration before flushing to the sink. If a changefeed pauses and then resumes, the `min_checkpoint_frequency` duration is the amount of time that the changefeed will need to catch up when it has not checkpointed its progress. During this catch-up time, you could receive duplicate messages.

Reviewer 2: ```suggestion
The `min_checkpoint_frequency` option controls how often nodes flush their progress to the coordinating changefeed node. Therefore, changefeeds will wait for at least the `min_checkpoint_frequency` duration before flushing to the sink. If a changefeed pauses and then resumes, the `min_checkpoint_frequency` duration is the amount of time that the changefeed will need to catch up since its previous checkpoint. During this catch-up time, you could receive duplicate messages.
```

nit: vale is throwing an error for checkpointed
